Apps/Src/Task_FaultState.c

```diff
@@ -1,34 +1,65 @@
 #include "os.h"
 #include "Tasks.h"
+#include "BSP_Contactor.h"
+#include "BSP_Lights.h"
+#include "Voltage.h"
+#include "Temperature.h"
+#include "Current.h"
+#include "EEPROM.h"
+#include "BSP_WDTimer.h"
+#include "CLI.h"
+#include "CANbus.h"
+#include "BSP_UART.h"
+#include "config.h"
+//Manthan Upadhyaya wuz here: 10/2020
 
 void Task_FaultState(void *p_arg) {
     (void)p_arg;
-
     OS_ERR err;
     CPU_TS ts;
 
     // BLOCKING =====================
     // Wait until a FAULT is signaled by another task.
-    OSSemPend(&Fault_Sem4,
-                0,
-                OS_OPT_PEND_BLOCKING,
-                &ts,
-                &err);
-
+    OSSemPend(&Fault_Sem4, 0, OS_OPT_PEND_BLOCKING, &ts, &err);
+    OSSemPost(&Fault_Sem4, OS_OPT_PEND_BLOCKING, &err);
     // Turn Contactor Off
-
+    BSP_Contactor_Off();
     // Turn Strobe Light On
-
-    // Turn LEDs On
-
-    // Record State of BPS into EEPROM
-
+    // Turn LEDs On and logs Error into EEPROM
+    BSP_Light_Off(RUN); //turn of run light
+    BSP_Light_On(FAULT);
+    if (Voltage_CheckStatus() == OVERVOLTAGE){
+        BSP_Light_On(OVOLT);
+        EEPROM_LogError(FAULT_HIGH_VOLT);
+    }
+    if (Voltage_CheckStatus() == UNDERVOLTAGE){
+        BSP_Light_On(UVOLT);
+        EEPROM_LogError(FAULT_LOW_VOLT);
+    }
+    if (Temperature_CheckStatus(Current_IsCharging()) == DANGER){
+        BSP_Light_On(OTEMP);
+        EEPROM_LogError(FAULT_HIGH_TEMP);
+    }
+    if (Current_CheckStatus(false) == DANGER){
+        BSP_Light_On(OCURR);
+        EEPROM_LogError(FAULT_HIGH_CURRENT);
+    }
+    if (BSP_WDTimer_DidSystemReset()){
+        BSP_Light_On(WDOG);
+        EEPROM_LogError(FAULT_WATCHDOG);
+    }
+    CANData_t data;
+    data.b = 1;
+    CANPayload_t Message;
+    Message.data = data;
     // Push Trip message to CAN Q
-
+    CANbus_Send(TRIP, Message);
     // Push Contactor State message to CAN Q
-
+    data.b = 0;
+    CANbus_Send(CONTACTOR_STATE, Message);
+    char command[COMMAND_SIZE];
     while(1) {
-        // CLI
-        // WDOG Reset
+        if (BSP_UART_ReadLine(command)) CLI_Handler(command); // CLI
```

Comment:
Can you explain why we need CLI in two tasks? I'm not saying this code is incorrect, this was just a question I had.

Reply:
This is the only way I could think of running the CLI task and reset the watchdog at the same time since the CLI task doesn't do that automatically.

Reply:
Why do we have to run the CLI in this task?

Reply:
I'd assume it's so we can read the data that led to our fault state. There might be other reasons too, but that's probably the main one.

Reply:
We want to run the CLI task here so we can figure out what caused the fault condition. I think this is fine here because if we get stuck in the CLI because of mutexes, the watchdog timer will trip and clear all of the mutexes, and we will return to a fault condition, so the contactor will stay open. What do you think @chaseblock @SijWoo @sugamarora99 @manthanand ?

Reply:
I discussed with @chaseblock over lunch, and I think it would be a good idea to add an `#ifdef` so that we only enable the CLI in the fault condition if we are debugging (I don't think this should be included when the car is racing because it can be very intrusive if it causes the watchdog timer to trip).